File: server/src/services/websocket.service.js
const { Server } = require('socket.io');
const jwt = require('jsonwebtoken');
const logger = require('../utils/logger');
const { getQueueSummary } = require('./queue.service');
const db = require('../db/pg_connection');

let io = null;

/**
 * Initialize Socket.IO server
 * @param {Object} httpServer - HTTP server instance
 */
function initializeWebSocket(httpServer) {
    io = new Server(httpServer, {
        cors: {
            origin: process.env.CORS_ORIGIN || '*',
            methods: ['GET', 'POST'],
            credentials: true,
        },
        pingTimeout: 60000,
        pingInterval: 25000,
    });

    // Authentication middleware
    io.use(async (socket, next) => {
        try {
            const token = socket.handshake.auth.token || socket.handshake.headers.authorization?.replace('Bearer ', '');

            if (!token) {
                return next(new Error('Authentication required'));
            }

            const decoded = jwt.verify(token, process.env.JWT_SECRET);
            socket.userId = decoded.id;
            socket.userRole = decoded.role;

            logger.info('WebSocket authenticated', { userId: decoded.id, role: decoded.role });
            next();
        } catch (err) {
            logger.error('WebSocket auth error', { error: err.message });
            next(new Error('Invalid token'));
        }
    });

    // Connection handler
    io.on('connection', (socket) => {
        logger.info('WebSocket connected', {
            socketId: socket.id,
            userId: socket.userId,
            role: socket.userRole,
        });

        // Join user's personal room
        socket.join(`user_${socket.userId}`);

        // Handle hospital room join (for staff)
        socket.on('join_hospital', async (hospitalId) => {
            if (['nurse', 'doctor', 'admin'].includes(socket.userRole)) {
                socket.join(`hospital_${hospitalId}`);
                logger.info('Joined hospital room', { userId: socket.userId, hospitalId });

                // Send initial queue summary
                try {
                    const summary = await getQueueSummary(hospitalId);
                    socket.emit('queue_summary', summary);
                } catch (err) {
                    logger.error('Error sending queue summary', { error: err.message });
                }
            }
        });

        // Handle ticket room join (for patients)
        socket.on('join_ticket', async (ticketId) => {
            try {
                const ticket = await db.findById('tickets', ticketId);

                // Verify ownership or staff access
                if (ticket.patient_id === socket.userId || ['nurse', 'doctor', 'admin'].includes(socket.userRole)) {
                    socket.join(`ticket_${ticketId}`);
                    logger.info('Joined ticket room', { userId: socket.userId, ticketId });

                    // Send current ticket status
                    socket.emit('ticket_status', ticket);
                } else {
                    socket.emit('error', { message: 'Access denied to this ticket' });
                }
            } catch (err) {
                logger.error('Error joining ticket room', { error: err.message });
                socket.emit('error', { message: 'Failed to join ticket room' });
            }
        });

        // Handle leave hospital
        socket.on('leave_hospital', (hospitalId) => {
            socket.leave(`hospital_${hospitalId}`);
            logger.info('Left hospital room', { userId: socket.userId, hospitalId });
        });

        // Handle leave ticket
        socket.on('leave_ticket', (ticketId) => {
            socket.leave(`ticket_${ticketId}`);
            logger.info('Left ticket room', { userId: socket.userId, ticketId });
        });

        // Disconnect handler
        socket.on('disconnect', (reason) => {
            logger.info('WebSocket disconnected', {
                socketId: socket.id,
                userId: socket.userId,
                reason,
            });
        });

        // Error handler
        socket.on('error', (err) => {
            logger.error('WebSocket error', {
                socketId: socket.id,
                userId: socket.userId,
                error: err.message,
            });
        });
    });

    logger.info('WebSocket server initialized');
    return io;
}

/**
 * Get Socket.IO instance
 */
function getIO() {
    if (!io) {
        throw new Error('Socket.IO not initialized');
    }
    return io;
}

/**
 * Emit queue update to hospital room
 */
async function emitQueueUpdate(hospitalId) {
    if (!io) return;

    try {
        const summary = await getQueueSummary(hospitalId);
        io.to(`hospital_${hospitalId}`).emit('queue_update', summary);
        logger.info('Queue update emitted', { hospitalId });
    } catch (err) {
        logger.error('Error emitting queue update', { error: err.message });
    }
}

/**
 * Emit ticket update to patient and hospital
 */
function emitTicketUpdate(ticket) {
    if (!io) return;

    try {
        // To patient
        io.to(`user_${ticket.patient_id}`).emit('ticket_update', ticket);
        io.to(`ticket_${ticket.id}`).emit('ticket_update', ticket);

        // To hospital staff
        io.to(`hospital_${ticket.hospital_id}`).emit('ticket_changed', ticket);

        logger.info('Ticket update emitted', { ticketId: ticket.id });
    } catch (err) {
        logger.error('Error emitting ticket update', { error: err.message });
    }
}

/**
 * Emit critical alert to hospital
 */
function emitCriticalAlert(hospitalId, ticket) {
    if (!io) return;

    try {
        io.to(`hospital_${hospitalId}`).emit('critical_alert', {
            ticketId: ticket.id,
            priority: ticket.priority_level,
            patientName: `${ticket.patient_prenom} ${ticket.patient_nom}`,
            timestamp: new Date().toISOString(),
        });

        logger.warn('Critical alert emitted', { hospitalId, ticketId: ticket.id });
    } catch (err) {
        logger.error('Error emitting critical alert', { error: err.message });
    }
}

/**
 * Emit notification to specific user
 */
function emitNotification(userId, notification) {
    if (!io) return;

    try {
        io.to(`user_${userId}`).emit('notification', notification);
        logger.info('Notification emitted', { userId, type: notification.type });
    } catch (err) {
        logger.error('Error emitting notification', { error: err.message });
    }
}

module.exports = {
    initializeWebSocket,
    getIO,
    emitQueueUpdate,
    emitTicketUpdate,
    emitCriticalAlert,
    emitNotification,
};

File: allo_urgence/lib/services/websocket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'package:flutter/foundation.dart';
import '../models/ticket.dart';

class WebSocketService with ChangeNotifier {
  IO.Socket? _socket;
  bool _isConnected = false;
  String? _error;

  // Getters
  bool get isConnected => _isConnected;
  String? get error => _error;

  /// Connect to WebSocket server
  void connect(String baseUrl, String token) {
    try {
      _socket = IO.io(
        baseUrl,
        IO.OptionBuilder()
            .setTransports(['websocket'])
            .disableAutoConnect()
            .setAuth({'token': token})
            .build(),
      );

      _socket!.onConnect((_) {
        debugPrint('✅ WebSocket connected');
        _isConnected = true;
        _error = null;
        notifyListeners();
      });

      _socket!.onDisconnect((_) {
        debugPrint('❌ WebSocket disconnected');
        _isConnected = false;
        notifyListeners();
      });

      _socket!.onConnectError((error) {
        debugPrint('❌ WebSocket connection error: $error');
        _error = error.toString();
        _isConnected = false;
        notifyListeners();
      });

      _socket!.onError((error) {
        debugPrint('❌ WebSocket error: $error');
        _error = error.toString();
        notifyListeners();
      });

      _socket!.connect();
    } catch (e) {
      debugPrint('❌ WebSocket setup error: $e');
      _error = e.toString();
      notifyListeners();
    }
  }

  /// Disconnect from WebSocket server
  void disconnect() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
    _isConnected = false;
    notifyListeners();
  }

  /// Join a ticket room to receive updates
  void joinTicket(String ticketId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('join_ticket', ticketId);
      debugPrint('📍 Joined ticket room: $ticketId');
    }
  }

  /// Leave a ticket room
  void leaveTicket(String ticketId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('leave_ticket', ticketId);
      debugPrint('📍 Left ticket room: $ticketId');
    }
  }

  /// Join a hospital room (for staff)
  void joinHospital(String hospitalId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('join_hospital', hospitalId);
      debugPrint('📍 Joined hospital room: $hospitalId');
    }
  }

  /// Leave a hospital room
  void leaveHospital(String hospitalId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('leave_hospital', hospitalId);
      debugPrint('📍 Left hospital room: $hospitalId');
    }
  }

  /// Listen to ticket updates
  void onTicketUpdate(Function(Map<String, dynamic>) callback) {
    _socket?.on('ticket_update', (data) {
      debugPrint('📊 Ticket update received: $data');
      callback(data as Map<String, dynamic>);
    });
  }

  /// Listen to queue updates
  void onQueueUpdate(Function(Map<String, dynamic>) callback) {
    _socket?.on('queue_update', (data) {
      debugPrint('📊 Queue update received: $data');
      callback(data as Map<String, dynamic>);
    });
  }

  /// Listen to critical alerts
  void onCriticalAlert(Function(Map<String, dynamic>) callback) {
    _socket?.on('critical_alert', (data) {
      debugPrint('🚨 Critical alert received: $data');
      callback(data as Map<String, dynamic>);
    });
  }

  /// Listen to notifications
  void onNotification(Function(Map<String, dynamic>) callback) {
    _socket?.on('notification', (data) {
      debugPrint('🔔 Notification received: $data');
      callback(data as Map<String, dynamic>);
    });
  }

  /// Remove all listeners
  void removeAllListeners() {
    _socket?.off('ticket_update');
    _socket?.off('queue_update');
    _socket?.off('critical_alert');
    _socket?.off('notification');
  }
}
